src/passes/GUFA.cpp

```diff
@@ -0,0 +1,319 @@
+      // TODO: tuple types.
+      return;
+    }
+
+    if (type.isRef() && (getTypeSystem() != TypeSystem::Nominal &&
+                         getTypeSystem() != TypeSystem::Isorecursive)) {
+      // Without type info we can't analyze subtypes, so we cannot infer
+      // anything about refs.
+      return;
+    }
+
+    // Ok, this is an interesting location that we might optimize. See what the
+    // oracle says is possible there.
+    auto contents = oracle.getContents(ExpressionLocation{curr, 0});
+
+    auto& options = getPassOptions();
+    auto& wasm = *getModule();
+    Builder builder(wasm);
+
+    auto replaceWithUnreachable = [&]() {
+      if (canReplaceWithUnreachable(curr)) {
+        replaceCurrent(getDroppedChildrenAndAppend(
+          curr, wasm, options, builder.makeUnreachable()));
+      } else {
+        // We can't remove this, but we can at least put an unreachable
+        // right after it.
+        replaceCurrent(builder.makeSequence(builder.makeDrop(curr),
+                                            builder.makeUnreachable()));
+      }
+      optimized = true;
+    };
+
+    if (contents.getType() == Type::unreachable) {
+      // This cannot contain any possible value at all. It must be unreachable
+      // code.
+      replaceWithUnreachable();
+      return;
+    }
+
+    // This is reachable. Check if we can emit something optimized for it.
+    // TODO: can we handle more general things here too?
+    if (!contents.canMakeExpression()) {
+      return;
+    }
+
+    if (contents.isNull() && curr->type.isNullable()) {
+      // Null values are all identical, so just fix up the type here if we need
+      // to (the null's type might not fit in this expression, if it passed
+      // through casts).
+      if (!Type::isSubType(contents.getType(), curr->type)) {
+        contents = PossibleContents::literal(
+          Literal::makeNull(curr->type.getHeapType()));
+      }
+
+      // Note that if curr's type is *not* nullable, then the code will trap at
+      // runtime (the null must arrive through a cast that will trap). We handle
+      // that below, so we don't need to think about it here.
+
+      // TODO: would emitting a more specific null be useful when valid?
+    }
+
+    auto* c = contents.makeExpression(wasm);
+
+    // We can only place the constant value here if it has the right type. For
+    // example, a block may return (ref any), that is, not allow a null, but in
+    // practice only a null may flow there if it goes through casts that will
+    // trap at runtime.
+    // TODO: GUFA should eventually do this, but it will require it properly
+    //       filtering content not just on ref.cast as it does now, but also
+    //       ref.as etc. Once it does those we could assert on the type being
+    //       valid here.
+    if (Type::isSubType(c->type, curr->type)) {
+      if (canRemove(curr)) {
+        replaceCurrent(getDroppedChildrenAndAppend(curr, wasm, options, c));
```

Comment:
Why can't we remove its children too? This is within `canRemove`, which checks `EffectAnalyzer` to see if it has any side effects. So that we are here means `curr` can be removed safely, no? The same question for this function's use in `replaceWithUnreachable`.

Reply:
Very good point! Yes, this could be a lot better. I fixed it to use `ShallowEffectAnalyzer` which is all we need since we'll keep the children if we actually need them. This optimizes more code + it's faster.

Reply:
I think using `ShallowEffectAnalyzer` is better, but what I was asking about was, why do we even need to keep children if `EffectAnalyzer.hasUnremovableSideEffects()` returns `false`? 

Some expression as a whole, including children, may not have a side effect even when some of children has one. For example, 
```wast
(block $lable0
  (br $label0)
)
```
`(br $label0)` has a side effects, but the whole `(block ... )` doesn't. I guess we don't encounter this case in this pass because we don't remove named blocks anyway, but actually in this case this block may be removable, because this `block` as a whole does not have any side effects.

So, if we are to preserve children anyway, using `ShallowEffectAnalyer` is better for sure. But in the previous code, when we were using `EffectAnalyer`, if that doesn't have any side effects, can't we remove the expression, including children, altogether? Do we even need to preserve children in that case as well?

Reply:
Yes, you're right that the earlier code was silly, since if `EffectAnalyzer.hasUnremovableSideEffects()` returns `false` then we don't need the children.

The new code seems optimal: we check for shallow effects which ignores the children, which allows us to remove more things, and then we do keep the children (if we need them).

Reply:
Isn't it more optimal if we don't keep children, because we can remove more code that way? I'm not sure why we need to keep children when `EffectAnalyzer.hasUnremovableSideEffects` returns false, which means the expression as a whole, including all children, doesn't have any side effects.

Reply:
> so the new code handles more cases, without losing anything (anything but the eager removal).

Is that always the case? There are cases when a child has a side effect but the parent, including all children, may not. Like the case I wrote above:
```wast
(block $lable0
  (br $label0)
)
```

Also `try`-`delegate` is a similar case; a single `try`-`delegate` may have a side effect (because `delegate` can target an outer `try`) but the outer `try` as a whole may not have any side effects.

But yeah, these may not matter because we exclude `block`s and `try`s in `canRemoveStructurally` or somewhere. But I wasn't sure it's always the case that "a child is not needed it will be removed by Vacuum eventually". Maybe it's true except the case of `block` and `try`, which we handle separately.

I think using `ShallowEffectAnalyzer` is better in terms of speed, because it's not O(n^2).

Reply:
Oh, right! Sorry for not understanding you before. I forgot about the possibility of the combination having fewer effects.

Yes, I think you're right. In principle this is an issue. However, `canRemoveStructurally` rules those cases out as you said. I think any time that the combined expression has no effects, but the pieces of it do if they are separate, are cases that we can't remove other pieces.

I added a comment here.

Reply:
OK now I think I understand better. Then what's the difference between the ones we exclude in `canRemoveStructurally` and the ones in `getDroppedUnconditionalChildrenAndAppend`?

In `canRemoveStructurally`, we exclude named `block`s, `try`s, and `pop`s. In `getDroppedUnconditionalChildrenAndAppend`, we exclude `if`s and `try`s. But I think the thing they check should be the same, namely, they check whether the expression itself can be removed, regardless of its chlidren can be removed or not, which will can be dropped separately.

So my question is, 1. Are they really the same? 2. If so, do we need two places to check this? 3. And, doesn't `getDroppedUnconditionalChildrenAndAppend` need to exclude `pop` and named `block`s as well, in case it is used in places other than GUFA? So what I mean is, can we check all these within `getDroppedUnconditionalChildrenAndAppend` and remove `canRemoveStructurally` altogether? We are currently excluding expressions in three places (`canRemove`, `canRemoveStructurally`, and `getDroppedUnconditionalChildrenAndAppend`), which is confusing. If we can remove one of them that will be an improvement.

Reply:
Hmm, I think this can actually be simplified a lot more, but it may be better to do that as a follow-up, given that this PR was hold up for too many weeks due to my laziness 😅 Sorry for that. I tried to do the follow-up in #4787.